service/collector/app_operator.go

```diff
@@ -0,0 +1,173 @@
+package collector
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/giantswarm/k8sclient/v3/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/prometheus/client_golang/prometheus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/app-operator/pkg/label"
+	"github.com/giantswarm/app-operator/pkg/project"
+)
+
+var (
+	appOperatorDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "ready", "total"),
+		"Gauge with ready app-operator instances per app CR version.",
+		[]string{
+			labelVersion,
+		},
+		nil,
+	)
+)
+
+// AppOperatorConfig is this collector's configuration struct.
+type AppOperatorConfig struct {
+	K8sClient k8sclient.Interface
+	Logger    micrologger.Logger
+}
+
+// AppOperator is the main struct for this collector.
+type AppOperator struct {
+	k8sClient k8sclient.Interface
+	logger    micrologger.Logger
+}
+
+// NewAppOperator creates a new AppOperator metrics collector
+func NewAppOperator(config AppOperatorConfig) (*AppOperator, error) {
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	a := &AppOperator{
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+	}
+
+	return a, nil
+}
+
+// Collect is the main metrics collection function.
+func (a *AppOperator) Collect(ch chan<- prometheus.Metric) error {
+	ctx := context.Background()
+
+	err := a.collectAppOperatorStatus(ctx, ch)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+// Describe emits the description for the metrics collected here.
+func (a *AppOperator) Describe(ch chan<- *prometheus.Desc) error {
+	ch <- appOperatorDesc
+	return nil
+}
+
+func (a *AppOperator) collectAppOperatorStatus(ctx context.Context, ch chan<- prometheus.Metric) error {
+	var err error
+
+	appVersions, err := a.collectAppVersions(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	operatorVersions, err := a.collectOperatorVersions(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for version := range appVersions {
+		var ready int32
+
+		if version == project.AppTenantVersion() {
+			// There should be a single app-operator instance with major version
+			// 1 for Helm 2 tenant clusters.
+			ready, err = helm2AppOperatorExists(operatorVersions)
```

Comment:
This part is a bit tricky. For Helm 2 there should only be a single instance with major version 1.